multierror.go

```diff
@@ -49,3 +50,69 @@ func (e *Error) GoString() string {
 func (e *Error) WrappedErrors() []error {
 	return e.Errors
 }
+
+// Unwrap returns an error from Error (or nil if there are no errors).
+// This error returned will further support Unwrap to get the next error,
+// etc. The order will match the order of Errors in the multierror.Error
+// at the time of calling.
+//
+// The resulting error supports errors.As/Is/Unwrap so you can continue
+// to use the stdlib errors package to introspect further.
+//
+// This will perform a shallow copy of the errors slice. Any errors appended
+// to this error after calling Unwrap will not be available until a new
+// Unwrap is called on the multierror.Error.
+func (e *Error) Unwrap() error {
+	// If we have no errors then we do nothing
+	if e == nil || len(e.Errors) == 0 {
+		return nil
+	}
+
+	// If we have exactly one error, we can just return that directly.
+	if len(e.Errors) == 1 {
+		return e.Errors[0]
+	}
+
+	// Shallow copy the slice
+	errs := make([]error, len(e.Errors))
```

Comment:
FWIW (caveat: I might be entirely misinterpreting the motivation for this code!) I think might be closing the stable door after the horse has bolted. If someone has returned `e` as an error somewhere else, then this might be called concurrently and you'll have a race condition regardless of whether you copy the slice here. ISTM that you should probably either make `Error` an opaque type with unexported fields and a constructor, or just drop this logic and just document that the Errors slice shouldn't be mutated after it's been returned as an `error` value.

Reply:
I'd be tempted to do the latter. That way you don't have any of the performance overhead of copying the slice every time the error is inspected.